src/coprocessor/dag/batch/executors/slow_hash_aggr_executor.rs

```diff
@@ -0,0 +1,401 @@
+    states_start_offset: usize,
+
+    /// The offset of each GROUP BY column in the group key. There will be an additional offset
+    /// at the end, which equals to the group key length.
+    ///
+    /// For example, if the group key `AaaFfffff` is composed by two group column `[Aaa, Ffffff]`,
+    /// then the value of this field is `[0, 3, 9]`.
+    ///
+    /// Note 1:
+    /// We want aggregation executor to provide each GROUP BY column independently, which will be
+    /// useful in future executors (like Projection executor). However, currently this segmentation
+    /// is not useful because we return data by row and columns will be always pasted together.
+    /// Thus segmentation or not doesn't make any differences.
+    ///
+    /// Note 2:
+    /// We use `SmallVec<[..; 6]>` so that when group by columns <= 5 SmallVec avoids allocation.
+    /// We think the case that there are more than 5 group by columns are rare.
+    ///
+    /// Note 3:
+    /// We use `SmallVec<[u32; ..]>` instead of `SmallVec<[usize; ..]>` because the length of the
+    /// group key will not exceed `u32`. Using `u32` instead of `usize` makes this field small
+    /// and can be fit into the cache line better.
+    group_key_offsets: SmallVec<[u32; 6]>,
+}
+
+impl<Src: BatchExecutor> AggregationExecutorImpl<Src> for SlowHashAggregationImpl {
+    #[inline]
+    fn prepare_entities(&mut self, entities: &mut Entities<Src>) {
+        let src_schema = entities.src.schema();
+        for group_by_exp in &self.group_by_exps {
+            entities
+                .schema
+                .push(group_by_exp.ret_field_type(src_schema).clone());
+        }
+    }
+
+    #[inline]
+    fn process_batch_input(
+        &mut self,
+        entities: &mut Entities<Src>,
+        mut input: LazyBatchColumnVec,
+    ) -> Result<()> {
+        let rows_len = input.rows_len();
+
+        // 1. Calculate which group each src row belongs to.
+        self.states_offset_each_row.clear();
+
+        let src_schema = entities.src.schema();
+
+        // TODO: Eliminate these allocations.
+        let mut group_by_keys = Vec::with_capacity(rows_len);
+        let mut group_by_keys_offsets = Vec::with_capacity(rows_len);
+        for _ in 0..rows_len {
+            group_by_keys.push(Vec::with_capacity(32));
+            group_by_keys_offsets.push(SmallVec::new());
+        }
+        for group_by_exp in &self.group_by_exps {
+            let group_by_result =
+                group_by_exp.eval(&mut entities.context, rows_len, src_schema, &mut input)?;
+            for row_index in 0..rows_len {
+                // Unwrap is fine because we have verified the group by expression before.
+                let group_column = group_by_result.vector_value().unwrap();
+                group_by_keys_offsets[row_index].push(group_by_keys[row_index].len() as u32);
+                group_column.encode(
+                    row_index,
+                    group_by_result.field_type(),
```

Comment:
Move this line out the loop too. I think compiler can inline this call but won't move out.